New optional parameters for course blocks API: lti_url, block_types

lenacom opened the pull request:
**Background:** Business people need to generate a list of LTI URLs for a given course, to provide to a customer. This new request should reduce the time and effort needed to pull LTI endpoints for a course, and reduce the chance of human error. 

**Studio Updates:** None.

**LMS Updates:** 2 new optional parameters are added to /api/courses/v1/blocks/?course_id=<course_id>: 
- 'requested_fields' can accept 'lti_url' (indicates it's required to show the field 'lti_url');
- 'block_types' (a list of requested block types; possible values include sequential, vertical, html, problem, video, and discussion). 

**Manual test instructions:**
Open in the browser:
http://127.0.0.1:8000/api/courses/v1/blocks/?course_id=course-v1%3AedX%2BDemoX%2BDemo_Course&username=staff&depth=all&block_counts=video&block_types=html,video&requested_fields=lti_url
![edx23](https://cloud.githubusercontent.com/assets/7777978/12007611/bb5b91d4-ac1c-11e5-8eb3-775f2dedd084.jpg)

edx-webhook commented:
Thanks for the pull request, @lenacom! I've created [OSPR-1025](https://openedx.atlassian.net/browse/OSPR-1025) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



You haven't added yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo. Please create a new pull request to do so!     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

lenacom commented:
jenkins run python

lenacom commented:
jenkins run quality

lenacom commented:
jenkins run js

singingwolfboy commented:
@lenacom, it looks like the tests that you added are failing on Jenkins. Can you address that, please?

lenacom commented:
All tests passed and the new request is verified manually after squashing commits.

nasthagiri commented:
Hi @lenacom,

@robrap pointed me to this PR and here's my initial feedback:

1. The course_structure API is actually being deprecated in favor of the new [Course API](https://github.com/edx/edx-platform/tree/master/lms/djangoapps/course_api), including the [Course Blocks API](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/course_api/blocks/views.py#L187) for retrieving all the blocks.

2. Instead of creating a separate endpoint for this, I would suggest adding an optional parameter to the Course Blocks API such as `lti=true`, which would then return lti-specific additional data, including the LTI URLs.

3. Lastly, if the end-users of this are "business people", then an API enhancement doesn't seem like a good interface.  For example, regardless of them needing to know how to make an API query, how would they know which block corresponds to a block they see visually on screen?  Have you considered simply adding a "Share via LTI" button in Studio or LMS that presents the URL to them?

dpowell commented:
Using the Course Blocks API makes sense to me, would you prefer a new Boolean optional parameter `include_lti_fields` over just adding the `lti_url` field to the existing `requested_fields` list?

nasthagiri commented:
Yes, adding it to `requested_fields` makes sense.  Can we be terse here and just call it `LTI`?

So the query would be:
```
/api/courses/v1/blocks/<usage_id>/?username=anjali&requested_fields=lti
```

dpowell commented:
Works for me.  We'll adjust this and resubmit shortly.

lenacom commented:
jenkins run lettuce

lenacom commented:
jenkins run python

jbarciauskas commented:
@nasthagiri @lenacom @dpowell what's the next step here?

dpowell commented:
@lenacom has made the changes @nasthagiri requested to use the new course APIs.  I believe we are waiting for another review?

robrap commented on lms/djangoapps/course_api/blocks/api.py:
```diff
@@ -17,7 +17,9 @@ def get_blocks(
         student_view_data=None,
-        return_type='dict'
+        return_type='dict',
+        block_types=None,
```
@nasthagiri When the api is called with block_counts filtered by certain block types, did you still want the rest of the blocks without the counts?

It looks like this is introducing a block type filter for all of the results, and could be in conflict with the block_counts block types.  What is best?:
1. Introduce this new overall filter and ignore/error on block types included in block_counts, but not included in block_types?
2. Always include all block types, and add block_lti_urls to match block_counts for how to turn on/off fields for specific block types?
3. Introduce new over filter and change block_counts to block_count True/False for including.  I'm guessing it's too late for this, and I'm not sure if it's what is wanted.

Thoughts?

nasthagiri replied:
The `block_types` parameter is a filter.  To be consistent with other filters implemented in this API, we should introduce a new `BlockTypesFilterTransformer`.  The new transformer would then be processed along with other Blocks API transformers, as done [here](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/course_api/blocks/transformers/blocks_api.py#L60).

Regarding conflicting with `block_counts`, I don't see it that way.  `block_counts` returns an aggregate count of the number of blocks within each subtree.  Whereas `block_types` is a filter, specifying exactly which types of blocks to return.  It's possible that `block_counts` may contain additional types that are not included in `block_types`.  In that case, the counts for those should still be returned, but the blocks themselves wouldn't.

In order to implement it as such, we would order the `block_counts` Transformer *before* the `block_types` Transformer.  Since the latter would filter out blocks before they are counted.

nasthagiri replied:
This API parameter list is getting unmanageably long.  Although out of the scope of this PR, we should take note to refactor this.

robrap replied:
Makes sense in terms of block_count. Thanks @nasthagiri. 
What did you have in mind for the potential refactor?

robrap commented on lms/djangoapps/course_api/blocks/api.py:
```diff
@@ -17,7 +17,9 @@ def get_blocks(
-        return_type='dict'
+        return_type='dict',
+        block_types=None,
+        lti_url=None
```
@nasthagiri Should we use something more like include_fields of proposed Course Discovery API, rather than adding these one add a time? https://openedx.atlassian.net/wiki/display/AC/Course+Discovery+API

robrap replied:
@lenacom @douglashall Would the LTI version ever affect these URLs? 

I'm still unclear about the differences between this and the endpoint documented here: https://github.com/edx/edx-platform/blob/a1c88bda5612f259b280776c68d06759ea2f75dc/lms/djangoapps/courseware/views.py#L1164-L1179
Maybe I could meet with you @douglashall to get cleared up?

douglashall replied:
@robrap As far as I can tell, the lti_url provided here would be what someone would use to add the block as an LTI provider to an LTI consumer.

The other endpoint you reference comes into play when edx is acting as the LTI consumer. When a student interacts with an edX LTI consumer component which has been added to a course, that LTI consumer component will include a URL in the LTI launch request sent to the LTI provider which the provider can use to pass grades back to the edx lms. The referenced endpoint above looks like it is intended to allow someone to get a list of all such URLs for a given course. Happy to step through it in person.

robrap replied:
Thanks @douglashall.  I still have some questions, and maybe the comment needs to be clarified in views.py, but I get that it is independent of this PR.  Thanks.  We'll meet soon.

nasthagiri replied:
@robrap @lenacom I am also surprised by the additional `ti_url` parameter.  I thought our earlier discussion indicated that we would simply add `lti` as an additional value to `requested_fields`.  So the request would be (copied from above):
```
/api/courses/v1/blocks/<usage_id>/?username=anjali&requested_fields=lti
```

robrap commented on lms/djangoapps/course_api/blocks/tests/test_api.py:
```diff
@@ -14,6 +14,7 @@ class TestGetBlocks(ModuleStoreTestCase):
     """
     Tests for the get_blocks function
     """
+
```
Don't make changes like this unless you are fixing a quality issue in the automated tests.

lenacom replied:
It's probably the result of autopep8

robrap commented on lms/djangoapps/course_structure_api/v0/urls.py:
```diff
@@ -18,7 +18,7 @@
         views.CourseGradingPolicy.as_view(),
         name='grading_policy'
-    ),
+    )
```
Don't make changes like this unless you are fixing a quality issue in the automated tests.  (Note that it was also inconsistent with the other one in this file.)

nedbat replied:
Trailing commas are a great feature of Python: they mean that new lines added after this one won't put this one into the next diff.  See earlier lines in this pull request for examples where you had to add a comma, and now there are two lines in the diff that aren't semantically interesting.

robrap commented on lms/djangoapps/course_structure_api/v0/views.py:
```diff
@@ -62,6 +62,7 @@ def course_check(func):
         :param func: function to be wrapped
         :returns: the wrapped function
         """
+
```
Don't make these changes (see whole file) unless you are fixing a quality issue in the automated tests.  Thanks.

robrap commented on lms/djangoapps/course_api/blocks/tests/test_serializers.py:
```diff
@@ -79,6 +83,7 @@ def assert_extended_block(self, serialized_block):
                 'id', 'type', 'lms_web_url', 'student_view_url',
                 'display_name', 'graded',
                 'block_counts', 'student_view_multi_device',
+                'lti_url'
```
I might just be missing it, but I don't see any test that the returned lti url is actually what you expect.

robrap commented on lms/djangoapps/course_api/blocks/tests/test_serializers.py:
```diff
@@ -40,6 +40,8 @@ def setUp(self):
             'block_structure': self.block_structure,
             'requested_fields': ['type'],
+            'lti_url': False,
+            'block_types': ['html']
```
I'd wait until the block_type discussion above gets resolved, but I don't see a test for when option block_types is not included.

robrap commented on lms/djangoapps/course_api/blocks/views.py:
```diff
@@ -85,6 +87,15 @@ class BlocksView(DeveloperErrorViewMixin, ListAPIView):
+
+          Example: block_types=vertical,html
+
+        * lti_url: (boolean) Indicates if it's required to show the field 'lti_url'.
```
Should follow other examples with optional/default details.  This may also need to wait on the discussion about how to add this to the api above (more generically, or one field at a time).

If this stayed, I'd change language to:
Indicates whether to include the field 'lti_url'.

nasthagiri commented:
@robrap @lenacom I am also surprised by the additional `ti_url` parameter.  I thought our earlier discussion indicated that we would simply add `lti` as an additional value to `requested_fields`.  So the request would be (copied from above):
```
/api/courses/v1/blocks/<usage_id>/?username=anjali&requested_fields=lti
```

nasthagiri commented on lms/djangoapps/course_api/blocks/serializers.py:
```diff
@@ -77,7 +85,10 @@ def get_blocks(self, structure):
+        for block_key in structure:
+            block_type = structure.get_xblock_field(block_key, 'category')
+            if not requested_block_types or block_type in requested_block_types:
+                result[unicode(block_key)] = BlockSerializer(block_key, context=self.context).data
```
This filtering shouldn't happen within the Serializer class.  Rather, please create a new Transformer for it (as noted in an earlier comment).

lenacom commented:
jenkins run lettuce

lenacom commented:
jenkins run a11y

lenacom commented:
jenkins run bokchoy

lenacom commented:
jenkins run python

lenacom commented:
jenkins run bokchoy

lenacom commented:
I moved 'lti_url' to the list in 'requested_fields' so that the link is now
http://127.0.0.1:8000/api/courses/v1/blocks/?course_id=course-v1%3AedX%2BDemoX%2BDemo_Course&username=staff&depth=all&block_counts=video&block_types=html,video&requested_fields=lti_url
Also removed not needed changes (empty lines, commas).

robrap commented:
@lenacom Can you let us know when the block type filtering is implemented using the transformers as detailed my @nasthagiri? Thanks.

lenacom commented:
@robrap I tried to follow the transformers pattern at first and encountered a problem. Blocks of requestes types can be unconnected so as a result we may need not an tree of blocks but a set of unconnected blocks. I don't know the code well, perhaps it's possible to implement this with transformers.

robrap commented:
Thanks @lenacom.  It's useful to know where you got caught up.

@nasthagiri Do you have an immediate answer to this or do you need to discuss?  Thanks.

nasthagiri commented:
@lenacom Thanks for looking into using the transformer design pattern to implement block-type filtering.  Yes, I can see your point about how a transformer (as it's currently designed) will not work for filtering out blocks - since the tree structure would be affected.

In that case, please do the filtering computation in api.py (as a separate private function).  I would move it out of serializer.py, since it's outside the scope of data serialization.

lenacom commented:
jenkins run lettuce

lenacom commented:
jenkins run bokchoy

lenacom commented:
@nasthagiri @robrap I implemented filtering by requested block types in a transformer using existing Block Structure API. There is one issue: a root block of type 'course' is always included in filtered results because it's required for a tree structure. It may be ok, we could just document this behaviour. What do you think?

robrap commented:
@lenacom I thought we were waiting on you. Sorry about that. Unfortunately I won't be able to get to this until early next week, but I didn't want to leave you with silence.  I'm not sure about @nasthagiri.

jbarciauskas commented:
@robrap will you still be able to take a look at this this week?

robrap commented:
@lenacom Thanks for your patience.  This Transformer code is going through a large refactor that @nasthagiri is working on.  We just had a nice design discussion, but we haven't finished it yet. 

The Transformer isn't going to work for filtering, so we need to handle that separately.  We'll provide more guidance soon.

robrap commented:
@lenacom I deleted my earlier comments temporarily because:
1. I made them against the commit by mistake.
2. I want you to hold off on changes until we give updated guidance.
Thank you!

robrap commented:
Hello @lenacom.  There was some refactoring around the transformer code, so this is going to need to be rebased.  Sorry about that.

It turns out that Transformers are not going to be the right solution for the Block Type filtering, because transforming happens at an earlier step that assumes pieces of the courses will not yet be filtered out.  

Although we have ideas for an extendable filtering system, like the Transformers, we don't have that yet.  @nasthagiri Has an idea for where the filtering code should live in the meantime.

Thanks again and sorry for the delay.

nasthagiri commented:
@lenacom Thanks for your patience.  The block-transformers currently require a hierarchical structure of the blocks.  This is in conflict with filters (such as block_types) that reduce the blocks to a set/list of blocks that are no longer connected in a tree/DAG structure.

So, I believe it makes sense to do the filtering outside of a transformer.  Eventually, we can imagine creating a filter-type transformer that doesn't expect blocks to be in a block structure.  If this is an undertaking you'd like to consider, then we can create a separate ticket and discuss the design for it further.  For example, we could create a new registry of filter-type transformers that expects a list of blocks.  And a new step in the Course Blocks API that iterates through all registered filter-type transformers.

As a short-term undertaking, you can do the following:
1. Update [BlockStructure.__iter__](https://github.com/edx/edx-platform/blob/master/openedx/core/lib/block_structure/block_structure.py#L73) to the following: `return self.get_block_keys`.
2. Apply your filter logic *after* the transformation step, but *before* the serialization step in [api.py](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/course_api/blocks/api.py#L64).

I believe that should be sufficient for your immediate needs.  I will be more responsive going forward on this PR, so please let me know if you have any further questions.

robrap commented:
@lenacom @dpowell Was this closed in favor of a different solution?  Is there anything else you need?  Thanks.

lenacom commented:
@robrap I'm rebasing the branch to master and removed all commits. I'm going to commit corrected implementation according to @nasthagiri instructions.

robrap commented:
Thanks for the update @lenacom.  If you can put a reference in the new PR to this one that would be great.
